### ISSUE TITLE
304 responses of conditional GETs are not handled properly

### ISSUE THREAD

**Issue by meshcow** [Azure/azure-storage-java#94]
CloduBlob's content download methods which accept OperationContext as a parameter  (such as `download()`, `downloadToByteArray()`, `downloadRangeToByteArray()`, `downloadRange()`, `downloadToFile()`) cannot handle NotModified (304) responses. For instance, a code like this
```java
CloudBlob blob = ... // get the blob's reference

OperationContext oc = new OperationContext();
Map<String, String> headers = new HashMap<>(1);
headers.put("if-none-match", blob.getProperties().getEtag());
oc.setUserHeaders(headers);

ByteArrayOutputStream os = new ByteArrayOutputStream(1);
blob.download(os, null, null, oc);
```
results in a ~2 minute delay after which the `download()` throws 
```
com.microsoft.azure.storage.StorageException: The condition specified using HTTP conditional header(s) is not met.
	at com.microsoft.azure.storage.StorageException.translateFromHttpStatus(StorageException.java:175)
	at com.microsoft.azure.storage.StorageException.translateException(StorageException.java:94)
	at com.microsoft.azure.storage.core.StorageRequest.materializeException(StorageRequest.java:305)
	at com.microsoft.azure.storage.core.ExecutionEngine.executeWithRetry(ExecutionEngine.java:175)
...
```

The response from the cloud, however, arrives right away, and is as follows:
```
HTTP/1.1 304 The condition specified using HTTP conditional header(s) is not met.
Content-Length: 0
Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
...
```
Would it be possible to improve the client side with proper handling logic for cases like this?

**Comment by esummers-msft** [comment]
Thank you for reporting this! It is a bug that we will fix in an upcoming release.

**Comment by meshcow** [comment]
Thanks! Looking forward to getting this fixed.

Meanwhile, I've discovered a more polite way of constructing the request without hacking with OperationContext :) It turned out there is a legal way to do so using AccessCondition.generateIfNoneMatchCondition(String etag). 
However, the result is the same.

**Comment by pemari-msft** [comment]
This is fixed as of our version 4.3.0 release. Please let us know if you have any further issues!
